File: _archived/documents-nodejs/src/domain/entities/document-version.ts
import type { ScanStatusValue } from '@/shared/constants';

/**
 * DocumentVersion entity — immutable once created.
 * Each upload of a new file creates a new version.
 */
export interface DocumentVersion {
  id:            string;
  documentId:    string;
  tenantId:      string;
  versionNumber: number;

  // Storage (internal only — never sent to clients)
  storageKey:    string;
  storageBucket: string;
  mimeType:      string;
  fileSizeBytes: number;
  checksum:      string;        // SHA-256

  // Scan lifecycle: PENDING → CLEAN | INFECTED | FAILED | SKIPPED
  scanStatus:        ScanStatusValue;
  scanCompletedAt:   Date | null;
  scanDurationMs:    number | null;
  scanThreats:       string[];        // populated when INFECTED
  scanEngineVersion: string | null;

  // Audit
  uploadedAt:    Date;
  uploadedBy:    string;
  label:         string | null;
  isDeleted:     boolean;
  deletedAt:     Date | null;
  deletedBy:     string | null;
}

export interface CreateVersionInput {
  documentId:    string;
  tenantId:      string;
  uploadedBy:    string;
  label?:        string;
  storageKey:    string;
  storageBucket: string;
  mimeType:      string;
  fileSizeBytes: number;
  checksum:      string;
  scanStatus:    ScanStatusValue;
  scanCompletedAt?: Date;
  scanDurationMs?:  number;
  scanThreats?:     string[];
  scanEngineVersion?: string;
}

File: apps/services/documents/Documents.Domain/Entities/DocumentVersion.cs
using Documents.Domain.Enums;

namespace Documents.Domain.Entities;

public sealed class DocumentVersion
{
    public Guid   Id               { get; set; }
    public Guid   DocumentId       { get; set; }
    public Guid   TenantId         { get; set; }
    public int    VersionNumber     { get; set; }
    public string MimeType         { get; set; } = string.Empty;
    public long   FileSizeBytes    { get; set; }

    // Storage — never exposed in API responses
    public string StorageKey       { get; set; } = string.Empty;
    public string StorageBucket    { get; set; } = string.Empty;
    public string? Checksum        { get; set; }

    // Scan
    public ScanStatus ScanStatus        { get; set; } = ScanStatus.Pending;
    public DateTime?  ScanCompletedAt   { get; set; }
    public int?       ScanDurationMs    { get; set; }
    public List<string> ScanThreats     { get; set; } = new();
    public string?    ScanEngineVersion { get; set; }

    // Metadata
    public string?  Label      { get; set; }
    public bool     IsDeleted  { get; set; }
    public DateTime? DeletedAt { get; set; }
    public Guid?    DeletedBy  { get; set; }

    // Audit
    public DateTime UploadedAt { get; set; }
    public Guid     UploadedBy { get; set; }

    // Navigation
    public Document Document { get; set; } = null!;
}
